v3/api/compute/instance_types.gen.go

```diff
@@ -0,0 +1,48 @@
+// Code generated by v3/generator; DO NOT EDIT.
+package compute
+
+import (
+	"context"
+
+	"github.com/exoscale/egoscale/v3/oapi"
+	"github.com/exoscale/egoscale/v3/utils"
+
+	openapi_types "github.com/deepmap/oapi-codegen/pkg/types"
+)
+
+type InstanceTypes struct {
+	oapiClient *oapi.ClientWithResponses
+}
+
+func NewInstanceTypes(c *oapi.ClientWithResponses) *InstanceTypes {
+	return &InstanceTypes{c}
+}
+
+func (a *InstanceTypes) List(ctx context.Context) ([]oapi.InstanceType, error) {
+	resp, err := a.oapiClient.ListInstanceTypesWithResponse(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	err = utils.ParseResponseError(resp.StatusCode(), resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return *resp.JSON200.InstanceTypes, nil
+}
+
+func (a *InstanceTypes) Get(ctx context.Context, id openapi_types.UUID) (*oapi.InstanceType, error) {
```

Comment:
In my opinion we shouldn't use types from the generated oapi code.

The fact that we generate code from an OpenAPI spec behind the scene and then generate a wrapper on top is a technical detail for the user, and they shouldn't see it.

Having all the types in the api/** packages would make better docs and, more importantly it would keep the user facing API contained, allowing us to keep it stable even if the technical details behind the scene are changing (deepmap/oapi-codegen upgrades, or even replacement if it doesn't fit our needs or isn't maintained anymore).

(If we had custom generation, it would be different though, because we would have more control on the generated code)

Reply:
Using types from oapi generated code is what makes generator very simple and wrapper thin. Copying structures to `api` namespace is doable, but i don't feel like it is worth the effort. I also don't think `oapi` code should be hidden, i see `api` and `oapi` as high level and low level api library that we offer.

Reply:
On this issue I tend to agree with @7fELF if we expose oapi types directly to users, we give up some control of our  API to `deepmap/oapi-codegen`. If we wrap the oapi types we will later be able to refuse changes that we don't agree with.
This even opens up the possibility of rewriting the code generation later. I still feel confident that our compromise of generating wrappers for `deepmap/oapi-codegen` instead of writing our own code generation tool makes sense. But if we make sure that the entire user-facing API is in our control, we keep the door open so that we may in the future rewrite the code generation process and remove the dependency on `deepmap/oapi-codegen`.

IIRC we are doing entire refactoring in part, because `deepmap/oapi-codegen` introduced breaking changes. IMHO avoiding this risk and having an API that we "fully" control justifies the extra effort of generating our own types.

Reply:
This was partially addressed in the latest iteration. Types are generated in the same (root) package and are now in the separate file.

We do not copy generated types, doing so is non-trivial task with little advantage. Since types are now separated from codegen functionality we can replace them independently if needed (which boils down to the same problem as if we maintain a copy of types).